File: Flatten Binary Tree to Linked List/Convert Sorted List to Binary Search Tree .py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return nothing, do it in place
    prev=None
    def recursive(self,root):
        if root==None:
            return
        else:
            self.prev.right=root
            self.prev.left=None
            self.prev=root
            rootr=root.right
            Solution.recursive(self,root.left)
            Solution.recursive(self,rootr)
    def flatten(self, root):
        newhead=TreeNode(0)
        newhead.right=root
        self.prev=newhead
        Solution.recursive(self,root)
   

File: Flatten Binary Tree to Linked List/Convert Sorted List to Binary Search Tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    // Keep track of previously visited node
private TreeNode prev = null;

/*
 * Recur version
 */
public void flatten(TreeNode root) {
    TreeNode dummy = new TreeNode(0);
    dummy.right = root;
    prev = dummy;
    flattenHelper(root);
}
public void flattenHelper(TreeNode root){
    if (root == null) return;
    prev.right = root;
    prev.left=null;   //add this line
    prev = root;
    // after invoke flattenHelper on root.left, root.right will revised
    // so record this value
    TreeNode rootRight = root.right;
    flattenHelper(root.left); 
    flattenHelper(rootRight); //
    }
}
